include/seqan3/core/copyable_wrapper.hpp

```diff
@@ -0,0 +1,248 @@
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::copyable_wrapper.
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <optional>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+/*!\brief Helper concept for seqan3::copyable_wrapper.
+ * \ingroup core
+ * \see https://en.cppreference.com/w/cpp/ranges/copyable_wrapper
+ * \noapi{Exposition only.}
+ */
+template <typename t>
+concept boxable = std::copy_constructible<t> && std::is_object_v<t>;
+
+/*!\brief Utility wrapper that behaves like std::optional but makes the type conform with the std::copyable concept.
+ * \ingroup core
+ * \see https://en.cppreference.com/w/cpp/ranges/copyable_wrapper
+ */
+template <typename t>
+class copyable_wrapper : public std::optional<t>
+{
+public:
+    using std::optional<t>::optional; //!< Use std::optional constructors.
+    using std::optional<t>::operator=; //!< Use std::optional assignment operators.
+    constexpr copyable_wrapper(copyable_wrapper const &) = default; //!< Defaulted.
+    constexpr copyable_wrapper(copyable_wrapper &&) = default; //!< Defaulted.
+    constexpr ~copyable_wrapper() = default; //!< Defaulted.
+
+    /*!\brief Use a specialised default constructor, if the wrapped type is default initialisable.
+     *        If not, the default constructor of std::optional is used.
+     */
+    constexpr copyable_wrapper() noexcept(std::is_nothrow_default_constructible_v<t>)
+        requires std::default_initializable<t>
+    : copyable_wrapper{std::in_place}
+    {}
+
+    //!\brief Copy assignment for non-copyable wrapped types.
+    constexpr copyable_wrapper & operator=(copyable_wrapper const & other)
+        noexcept(std::is_nothrow_copy_constructible_v<t>)
+        requires (!std::copyable<t>)
+    {
+        if (this != std::addressof(other))
+        {
+            if (other)
+                this->emplace(*other);
+            else
+                this->reset();
+        }
+        return *this;
+    }
+
+    //!\brief Move assignment for non-movable wrapped types.
+    constexpr copyable_wrapper & operator=(copyable_wrapper && other) noexcept(std::is_nothrow_move_constructible_v<t>)
+        requires (!std::movable<t>)
+    {
+        if (this != std::addressof(other))
+        {
+            if (other)
+                this->emplace(std::move(*other));
+            else
+                this->reset();
+        }
+        return *this;
+    }
+
+    /*!\brief Invokes the wrapped object with passed arguments.
+     * \throws std::bad_optional_access if no value is contained.
+     * \details
+     *
+     * This is a SeqAn-specific extension that allows easy invocation of the wrapped object.
+     * `t` needs to be callable with the passed arguments.
+     *
```

Comment:
Can you give a short example?

Reply:
```cpp
seqan3::copyable_wrapper wrapper{[] (int const x) { return x + 1; }};
auto wrapper_2 = wrapper; // A lambda function wouldn't be copyable.
wrapper(2); // Returns 3
wrapper_2(2); // Returns 3
```

It's a functionality that `semiregular_box` had, but is not part of the Standard.
And we use it quite heavily, the alternative would be `wrapper->operator()(2)`.

Reply:
```suggestion
     * This is a SeqAn-specific extension that allows easy invocation of the wrapped object.
     * `t` needs to be callable with the passed arguments.
     * e.g.:
     * ```
     * seqan3::copyable_wrapper wrapper{[] (int const x) { return x + 1; }};
     * auto wrapper_2 = wrapper; // A lambda function wouldn't be copyable.
     * wrapper(2); // Returns 3
     * wrapper_2(2); // Returns 3
     * ```
     *

since this is deep detail I'd be fine with no snippet but hard coded code